lib/Dialect/ModArith/Conversions/ModArithToArith/ModArithToArith.cpp

```diff
@@ -250,6 +269,127 @@ struct ConvertMac : public OpConversionPattern<MacOp> {
   }
 };
 
+struct ConvertModSwitch : public OpConversionPattern<ModSwitchOp> {
+  ConvertModSwitch(mlir::MLIRContext *context)
+      : OpConversionPattern<ModSwitchOp>(context) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  auto getInteger(ImplicitLocOpBuilder *b, IntegerType integerType) const {
+    return [=](const APInt &value) mutable {
+      return b->create<arith::ConstantOp>(IntegerAttr::get(integerType, value));
+    };
+  }
+
+  LogicalResult matchAndRewrite(
+      ModSwitchOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+    auto inputType = op.getInput().getType();
+    auto outputType = op.getOutput().getType();
+    if (auto inputModArith = dyn_cast<ModArithType>(inputType)) {
+      if (auto outputModArith = dyn_cast<ModArithType>(outputType)) {
+        auto oldModulus = inputModArith.getModulus().getValue();
+        auto newModulus = outputModArith.getModulus().getValue();
+        auto oldWidth = oldModulus.getBitWidth();
+        auto newWidth = newModulus.getBitWidth();
+        auto width = std::max(oldWidth, newWidth);
+        auto integerType = IntegerType::get(op.getContext(), width);
+        auto intConst = getInteger(&b, integerType);
+        oldModulus = oldModulus.zext(width);
+        newModulus = newModulus.zext(width);
+        Value input = adaptor.getInput();
+        if (newWidth > oldWidth) {
+          input = b.create<arith::ExtUIOp>(integerType, input);
+        }
+        auto cmod = intConst(newModulus);
+        auto remu = b.create<arith::RemUIOp>(input, cmod);
+        auto cmp = b.create<arith::CmpIOp>(arith::CmpIPredicate::ugt,
+                                           intConst(oldModulus.lshr(1)), input);
+        auto diffConst = intConst(newModulus - oldModulus.urem(newModulus));
+        auto remuAdd = b.create<arith::AddIOp>(remu, diffConst);
+        auto remuAddMod =
+            b.create<arith::RemUIOp>(remuAdd, intConst(newModulus));
+        Value result = b.create<arith::SelectOp>(cmp, remu, remuAddMod);
+        if (newWidth < oldWidth) {
+          auto outputType = IntegerType::get(op.getContext(), newWidth);
+          result = b.create<arith::TruncIOp>(outputType, result);
+        }
+        rewriter.replaceOp(op, result);
```

Comment:
I think this would be easier to read if either the different cases were separated into different functions, or else the nesting was reduced by returning early at the end of each block (with slightly more verbose conditions). E.g.,

```
if (auto inputModArith = dyn_cast<ModArithType>(inputType) && auto outputModArith = dyn_cast<ModArithType>(outputType)) {
   ...
   return success();
}

if (auto inputModArith = dyn_cast<ModArithType>(inputType) && auto outputRns = dyn_cast<rns::RNSType>(outputType)) {
   ...
   return success();
}

...
```